### PR TITLE
Rich notifications support

### LINKED ISSUE #268: Rich notifications enhancements
I had started the work to include extra functionality for rich notifications (serialization of the certificate, validation & decryption of payload) in the dotnet SDK back when I was a PM for the feature. Would you mind taking over my old PRs, rebasing and merging the work so it's part of the upcoming release?

https://github.com/microsoftgraph/msgraph-sdk-dotnet/pull/754
https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/117
https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/pull/104

_Originally posted by @baywet in https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/issues/107#issuecomment-860640130_ 
[AB#10062](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/10062)

### PR DESCRIPTION
**Pull request by andrueastman**
This PR supercedes #117 and closes #268.

It involves cleanup of work done in #117

Once merged, the SDK user is able to simplify the work of handling encryptions and tokens with subscriptions by doing the following.

1. Create a subscription
```cs
            // create a subscription to listen to new and edited teams messages sent 
            var subscription = new Subscription
            {
                ChangeType = "created,updated",
                IncludeResourceData = true,
                NotificationUrl = _config.Ngrok + "/api/notifications",
                Resource = "/teams/getAllMessages",
                ExpirationDateTime = DateTime.UtcNow.AddMinutes(5),
                ClientState = "SecretClientState",
                EncryptionCertificateId = "my-custom-id",
            };

            // Load a X509Certificate in to the subscription object.
            subscription.AddPublicEncryptionCertificate(this._certificate);

            var newSubscription = await graphServiceClient
              .Subscriptions
              .Request()
              .AddAsync(subscription);
```

2. Receiving notifications with encrypted data
```cs
        public async Task<ActionResult<string>> Post([FromQuery] string validationToken = null)
        {
            // handle validation
            if (!string.IsNullOrEmpty(validationToken))
            {
                Console.WriteLine($"Received Token: '{validationToken}'");
                return Ok(validationToken);
            }
            var graphServiceClient = GetGraphClient();
            var myTenantIds = new Guid[] { new Guid(_config.TenantId) };
            var myAppIds = new Guid[] { new Guid(_config.AppId) };
            
            // handle notifications
            var content = await new StreamReader(Request.Body).ReadToEndAsync();
            var collection = graphServiceClient.HttpProvider.Serializer.DeserializeObject<ChangeNotificationCollection>(content);
            var areTokensValid = await collection.AreTokensValid(myTenantIds, myAppIds);
            foreach (var changeNotification in collection.Value)
            {
                var attachedChatMessage = await changeNotification.EncryptedContent.DecryptAsync<ChatMessage>((id, thumbprint) => Task.FromResult(this._certificate));
                if (areTokensValid)
                {
                    // handle the decrypted object infromation
                    Console.WriteLine($"Message time: {attachedChatMessage.CreatedDateTime}");
                    Console.WriteLine($"Message from: {attachedChatMessage.From?.User?.DisplayName}");
                    Console.WriteLine($"Message content: {attachedChatMessage.Body.Content}");
                    Console.WriteLine();
                }
            }
            return Ok();
        }
```

Unit tests have been added to validate the functionality of the added extension methods.

### Related PRs to be merged after this
- (v1) https://github.com/microsoftgraph/msgraph-sdk-dotnet/pull/1053
- (beta) https://github.com/microsoftgraph/msgraph-beta-sdk-dotnet/pull/315

Note: The work on this PR can also be validated/tested locally with the test application [here](https://github.com/andrueastman/RichNotifications).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/269)

### CODE REVIEW

**Review by MIchaelMainer — APPROVED** [review]
Approved with suggestions. Vincent should review this too.

Suggestions:

1) You can remove indent levels and remove unneeded curly braces around using statements. https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations

2) Method calls into System.Security.Cryptography can result in exceptions. I suggest that you expose that possibility in an exception doc tag. https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments#exception. This way IDEs can show this developers and they can choose whether they want to handle these.

3) Long param code comment length. IDEs will make very long lines for these scenarios. Consider breaking the really long lines into two lines.

File: src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                    using (var memoryStream = new MemoryStream())
+                    using (var cryptoStream = new CryptoStream(memoryStream, cryptoServiceProvider.CreateDecryptor(), CryptoStreamMode.Write))
+                    {
+                        cryptoStream.Write(dataToDecrypt, 0, dataToDecrypt.Length);
+                        cryptoStream.FlushFinalBlock();
+                        return memoryStream.ToArray();
+                    }
+                }
+            }
+            catch (Exception ex)
```
[the remark this answers is not in the document]
Comment (andrueastman) [reply src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs:81]: Thanks for the feedback. These should be resolved via https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/269/commits/994df340207399a39c12375fbbbd7a177b0676ca

**Review by MIchaelMainer — DISMISSED** [review]

File: src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        /// Validates the signature and decrypted content attached with the notification.
+        /// </summary>
+        /// <typeparam name="T">Type to deserialize the data to.</typeparam>
+        /// <param name="encryptedContent">The encrypted content of type <see cref="IDecryptableContent"/></param>
+        /// <param name="certificateProvider">Certificate provider to decrypt the content.
+        /// The first parameter is the certificate ID provided when creating the subscription.
+        /// The second is the certificate thumbprint. The certificate WILL be disposed at the end of decryption.</param>
+        /// <returns>Decrypted content as the provided type.</returns>
+        public static async Task<T> DecryptAsync<T>(this IDecryptableContent encryptedContent, Func<string, string, Task<X509Certificate2>> certificateProvider) where T : class
+        {
```
Comment (baywet) [inline src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs:30]: defensive programing

File: src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        /// <param name="encryptedContent">The encrypted content of type <see cref="IDecryptableContent"/></param>
+        /// <param name="certificateProvider">Certificate provider to decrypt the content.
+        /// The first parameter is the certificate ID provided when creating the subscription.
+        /// The second is the certificate thumbprint. The certificate WILL be disposed at the end of decryption.</param>
+        /// <exception cref="InvalidDataException">Thrown when the <see cref="IDecryptableContent.DataSignature"/> value does not match the signature in the payload</exception>
+        /// <exception cref="ApplicationException">Thrown when there is a failure in attempting to decrypt the information</exception>
+        /// <returns>Decrypted content as string.</returns>
+        public static async Task<string> DecryptAsync(this IDecryptableContent encryptedContent, Func<string, string, Task<X509Certificate2>> certificateProvider)
+        {
+            using var certificate = await certificateProvider(encryptedContent.EncryptionCertificateId, encryptedContent.EncryptionCertificateThumbprint);
```
Comment (baywet) [inline src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs:47]: defensive programing

File: src/Microsoft.Graph.Core/Extensions/IEncryptableSubscriptionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    public static class IEncryptableSubscriptionExtensions
+    {
+        /// <summary>
+        /// Adds the public encryption certificate information for change notifications with resource data to the subscription creation information.
+        /// </summary>
+        /// <param name="subscription">The subscription instance of type <see cref="IEncryptableSubscription"/></param>
+        /// <param name="certificate">Certificate to use for encryption</param>
+        public static void AddPublicEncryptionCertificate(this IEncryptableSubscription subscription, X509Certificate2 certificate)
+        {
+            subscription.EncryptionCertificate = Convert.ToBase64String(certificate.Export(X509ContentType.Cert));
```
Comment (baywet) [inline src/Microsoft.Graph.Core/Extensions/IEncryptableSubscriptionExtensions.cs:22]: defensive programing

**Review by baywet — CHANGES_REQUESTED** [review]
thanks for picking this up! A few minor comments and I think we'll be good to merge it.

File: src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs (hunk 2 above)
Comment (andrueastman) [reply src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs:31]: Resolved via https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/269/commits/467c33678becdfa1093a184a37abc46251c7966e

File: src/Microsoft.Graph.Core/Extensions/IEncryptableSubscriptionExtensions.cs (hunk 4 above)
Comment (andrueastman) [reply src/Microsoft.Graph.Core/Extensions/IEncryptableSubscriptionExtensions.cs:26]: Resolved via https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/269/commits/467c33678becdfa1093a184a37abc46251c7966e

File: src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs (hunk 3 above)
Comment (andrueastman) [reply src/Microsoft.Graph.Core/Extensions/IDecryptableContentExtensions.cs:55]: Resolved via https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/269/commits/467c33678becdfa1093a184a37abc46251c7966e

**Review by baywet — APPROVED** [review]
Thanks for taking in the changes!

**Pull request merged**